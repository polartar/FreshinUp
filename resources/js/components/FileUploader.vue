<template>
  <v-layout
    column
    fill-height
  >
    <v-layout
      row
      class="subheading font-weight-bold"
    >
      <v-flex
        v-if="uploading"
      >
        <v-progress-linear v-model="uploadPercentage" />
        <div>{{ uploadPercentage }}%</div>
      </v-flex>
      <v-flex
        v-else
      >
        File:
        <a
          v-if="isDownloadable"
          flat
          :href="src"
          download
        >
          {{ name }}
        </a>
      </v-flex>
    </v-layout>
    <v-layout
      row
      wrap
      justify-space-between
    >
      <v-flex
        xs5
      >
        <v-btn
          block
          color="primary"
          @click="launchFilePicker"
        >
          Upload file
        </v-btn>
      </v-flex>
      <v-flex
        xs5
      >
        <v-btn
          block
          color="error"
          @click="remove"
        >
          Delete File
        </v-btn>
      </v-flex>
    </v-layout>

    <input
      v-show="false"
      ref="file"
      type="file"
      @change="onFileChange($event.target.files)"
    >

    <!-- error dialog displays any potential errors -->
    <v-dialog
      v-model="errorDialog"
      max-width="300"
    >
      <v-card>
        <v-card-text class="subheading">
          {{ errorText }}
        </v-card-text>
        <v-card-actions>
          <v-spacer />
          <v-btn
            flat
            @click="errorDialog = false"
          >
            Got it!
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </v-layout>
</template>

<script>
import axios from 'axios'
import MapValueKeysToData from '../mixins/MapValueKeysToData'
import get from 'lodash/get'

const MB = 1024 * 1024
export const DEFAULT_VALUE = {
  name: '',
  src: ''
}

/**
 * @property string name
 * @property string src
 */
export default {
  mixins: [MapValueKeysToData],
  model: {
    prop: 'value',
    event: 'onValueChange'
  },
  props: {
    value: { type: Object, default: () => DEFAULT_VALUE },
    maxFileSize: { type: Number, default: 100 }
  },
  data () {
    return {
      ...DEFAULT_VALUE,
      errorDialog: false,
      errorText: '',
      uploadPercentage: 0,
      uploading: false
    }
  },
  computed: {
    maxInBytes () {
      return MB * this.maxFileSize
    },
    isDownloadable () {
      return Boolean(this.src && this.name)
    }
  },
  methods: {
    launchFilePicker () {
      this.$refs.file.click()
    },
    async onFileChange (files) {
      const fileSource = get(files, '0')
      if (!fileSource) {
        this.showError(`No file selected.`)
        return false
      }
      const size = fileSource.size - this.maxInBytes
      if (size > 1) {
        this.showError(`Your file is too big! Please select a file under ${this.maxFileSize}MB`)
        return false
      }
      const filePath = await this.submitFile(fileSource)
      if (filePath) {
        this.src = filePath
        this.name = fileSource.name
        this.$emit('input', { name: fileSource.name, src: filePath })
      }
    },
    async submitFile (file) {
      try {
        this.uploading = true
        let formData = new FormData()
        formData.append('file', file)
        const response = await axios.post('/foodfleet/tmp-media',
          formData,
          {
            headers: {
              'Content-Type': 'multipart/form-data'
            },
            onUploadProgress: function (progressEvent) {
              this.uploadPercentage = parseInt(Math.round((progressEvent.loaded * 100) / progressEvent.total)) || 0
            }.bind(this)
          }
        )
        this.uploading = false
        this.uploadPercentage = 0
        return get(response, 'data')
      } catch (e) {
        this.uploading = false
        this.uploadPercentage = 0
        return false
      }
    },
    showError (text) {
      this.errorText = text
      this.errorDialog = true
    },
    remove () {
      this.$refs.file.value = ''
      this.$emit('input', DEFAULT_VALUE)
    }
  }
}
</script>
