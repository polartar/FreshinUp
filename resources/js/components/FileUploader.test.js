import { shallowMount, mount } from '@vue/test-utils'
import createLocalVue from 'vue-cli-plugin-freshinup-ui/utils/testing/createLocalVue'
import Component from './FileUploader.vue'
import * as Stories from './FileUploader.stories'

function MockFile (name, size, mimeType) {
  name = name || 'mock.txt'
  size = size || 1024
  mimeType = mimeType || 'application/pdf'

  function range (count) {
    return Array(count).fill(0).join('')
  }

  let blob = new Blob([range(size)], { type: mimeType })
  blob.lastModifiedDate = new Date()
  blob.name = name

  return blob
}

describe('components/FileUploader', () => {
  // Component instance "under test"
  let mock
  describe('Snapshots', () => {
    test('Default', () => {
      const wrapper = mount(Stories.Default())
      expect(wrapper.element).toMatchSnapshot()
    })
    test('Downloadable', () => {
      const wrapper = mount(Stories.Downloadable())
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('Props & Computed', () => {
    test('value', async () => {
      const wrapper = shallowMount(Component)
      expect(wrapper.vm.value).toMatchObject({
        name: '',
        src: ''
      })

      const value = {
        name: 'my file',
        src: 'https://example.domain/my-file.csv'
      }
      wrapper.setProps({
        value
      })
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.value).toMatchObject(value)
      expect(wrapper.vm.name).toEqual(value.name)
      expect(wrapper.vm.src).toEqual(value.src)
    })

    // computed
    test('maxInBytes', () => {
      const wrapper = shallowMount(Component)
      const MB = 1024 * 1024

      wrapper.setProps({
        maxFileSize: 5
      })
      expect(wrapper.vm.maxInBytes).toEqual(5 * MB)

      wrapper.setProps({
        maxFileSize: 10
      })
      expect(wrapper.vm.maxInBytes).toEqual(10 * MB)
    })

    test('isDownloadable', async () => {
      const wrapper = shallowMount(Component)
      expect(wrapper.vm.isDownloadable).toEqual(false)

      wrapper.setProps({
        value: { name: 'mock name', src: 'https://downloadable.net/mockfile' }
      })
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.isDownloadable).toEqual(true)
    })
  })

  describe('Methods', () => {
    beforeEach(() => {
      const vue = createLocalVue()
      mock = vue.mock
      mock
        .onPost('foodfleet/tmp-media').reply(200, 'mock url')
        .onAny().reply(config => {
          console.warn('No mock match for ' + config.url, config)
          return [404, {}]
        })
    })

    test('launchFilePicker()', () => {
      const wrapper = shallowMount(Component)
      const clickMock = jest.fn()
      wrapper.vm.$refs.file = {
        click: clickMock
      }
      wrapper.vm.launchFilePicker()
      expect(clickMock).toHaveBeenCalled()
    })

    describe('onFileChange()', () => {
      test('default', async () => {
        const wrapper = shallowMount(Component)
        const file = new MockFile()
        await wrapper.vm.onFileChange([file])
        const emitted = wrapper.emitted().input
        expect(emitted).toBeTruthy()
        expect(emitted[0][0].name).toEqual('mock.txt')
        expect(emitted[0][0].src).toEqual('mock url')
      })
      test('when more than max file size display error', () => {
        const wrapper = shallowMount(Component, {
          propsData: {
            value: { name: '', src: '' },
            maxFileSize: 1
          }
        })
        const file = new MockFile(null, 10 * 1024 * 1024, null)
        wrapper.vm.onFileChange([ file ])
        expect(wrapper.vm.errorDialog).toBeTruthy()
      })
    })

    test('submitFile()', async () => {
      const wrapper = shallowMount(Component)
      const file = new MockFile()
      const src = await wrapper.vm.submitFile([ file ])
      expect(src).toEqual('mock url')
    })

    test('remove()', () => {
      const wrapper = shallowMount(Component, {
        propsData: {
          value: { name: 'test_name', src: '' }
        }
      })
      wrapper.vm.remove()
      const emitted = wrapper.emitted().input
      expect(emitted).toBeTruthy()
      expect(emitted[0][0].name).toEqual('')
      expect(emitted[0][0].src).toEqual('')
    })

    test('showError(message)', () => {
      const wrapper = shallowMount(Component)
      const errorText = 'error text'
      wrapper.vm.showError(errorText)
      expect(wrapper.vm.errorText).toEqual(errorText)
      expect(wrapper.vm.errorDialog).toEqual(true)
    })
  })
})
