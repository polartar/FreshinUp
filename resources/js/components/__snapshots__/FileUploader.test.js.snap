// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FileUploader Snapshots Default 1`] = `
<div
  class="container"
>
  <div
    class="layout column fill-height"
  >
    <div
      class="layout subheading font-weight-bold row"
    >
      <div
        class="flex"
      >
        
      File:
      
        <!---->
      </div>
    </div>
     
    <div
      class="layout row wrap justify-space-between"
    >
      <div
        class="flex xs5"
      >
        <button
          class="v-btn v-btn--block theme--light primary"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            
        Upload file
      
          </div>
        </button>
      </div>
       
      <div
        class="flex xs5"
      >
        <button
          class="v-btn v-btn--block theme--light error"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            
        Delete File
      
          </div>
        </button>
      </div>
    </div>
     
    <input
      style="display: none;"
      type="file"
    />
     
    <div
      class="v-dialog__container"
      style="display: block;"
    >
      <div
        class="v-dialog__content"
        style="z-index: 0;"
        tabindex="-1"
      >
        <div
          class="v-dialog"
          style="max-width: 300px; display: none;"
        >
          <div
            class="v-card v-sheet theme--light"
          >
            <div
              class="v-card__text subheading"
            >
              
        
      
            </div>
             
            <div
              class="v-card__actions"
            >
              <div
                class="spacer"
              />
               
              <button
                class="v-btn v-btn--flat theme--light"
                type="button"
              >
                <div
                  class="v-btn__content"
                >
                  
          Got it!
        
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/FileUploader Snapshots Downloadable 1`] = `
<div
  class="container"
>
  <pre>
            {
  "name": "mock name",
  "src": "https://downloadable.net/mock"
}
      
  </pre>
   
  <div
    class="layout column fill-height"
  >
    <div
      class="layout subheading font-weight-bold row"
    >
      <div
        class="flex"
      >
        
      File:
      
        <!---->
      </div>
    </div>
     
    <div
      class="layout row wrap justify-space-between"
    >
      <div
        class="flex xs5"
      >
        <button
          class="v-btn v-btn--block theme--light primary"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            
        Upload file
      
          </div>
        </button>
      </div>
       
      <div
        class="flex xs5"
      >
        <button
          class="v-btn v-btn--block theme--light error"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            
        Delete File
      
          </div>
        </button>
      </div>
    </div>
     
    <input
      style="display: none;"
      type="file"
    />
     
    <div
      class="v-dialog__container"
      style="display: block;"
    >
      <div
        class="v-dialog__content"
        style="z-index: 0;"
        tabindex="-1"
      >
        <div
          class="v-dialog"
          style="max-width: 300px; display: none;"
        >
          <div
            class="v-card v-sheet theme--light"
          >
            <div
              class="v-card__text subheading"
            >
              
        
      
            </div>
             
            <div
              class="v-card__actions"
            >
              <div
                class="spacer"
              />
               
              <button
                class="v-btn v-btn--flat theme--light"
                type="button"
              >
                <div
                  class="v-btn__content"
                >
                  
          Got it!
        
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
